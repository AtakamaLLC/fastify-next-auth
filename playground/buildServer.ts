import Fastify from 'fastify'
import fastifyEnv from '@fastify/env'
import GithubProvider from 'next-auth/providers/github'
import NextAuthPlugin from 'fastify-next-auth'
import type { NextAuthOptions } from 'fastify-next-auth'

const schema = {
  type: 'object',
  required: ['NEXTAUTH_URL', 'NEXTAUTH_SECRET', 'GITHUB_CLIENT_ID', 'GITHUB_CLIENT_SECRET'],
  properties: {
    NEXTAUTH_URL: {
      type: 'string',
      default: 'http://localhost:3000',
    },
    NEXTAUTH_SECRET: {
      type: 'string',
    },
    GITHUB_CLIENT_ID: {
      type: 'string',
    },
    GITHUB_CLIENT_SECRET: {
      type: 'string',
    },
  },
}

export async function buildServer() {
  const server = Fastify()

  async function initialize() {
    server.register(fastifyEnv, {
      schema,
      dotenv: true,
    })
    await server.after()
    server.register(NextAuthPlugin, {
      secret: process.env.NEXTAUTH_SECRET,
      providers: [
        GithubProvider({
          clientId: process.env.GITHUB_CLIENT_ID,
          clientSecret: process.env.GITHUB_CLIENT_SECRET,
        }),
      ],
    } as NextAuthOptions)
  }

  initialize()

  async function startServer() {
    try {
      await server.ready()
      await server.listen({
        port: 3000,
      })
    }
    catch (err) {
      server.log.error(err)
      process.exit(1)
    }
  }

  startServer()

  return server
}
