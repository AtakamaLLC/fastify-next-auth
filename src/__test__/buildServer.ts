import path from 'path'
import Fastify from 'fastify'
import fastifyEnv from '@fastify/env'
import GithubProvider from 'next-auth/providers/github'
import NextAuthPlugin from '../index'
import type { NextAuthOptions } from '../index'

const schema = {
  type: 'object',
  required: ['NEXTAUTH_URL', 'NEXTAUTH_SECRET', 'GITHUB_CLIENT_ID', 'GITHUB_CLIENT_SECRET'],
  properties: {
    NEXTAUTH_URL: {
      type: 'string',
      default: 'http://localhost:3000',
    },
    NEXTAUTH_SECRET: {
      type: 'string',
    },
    GITHUB_CLIENT_ID: {
      type: 'string',
    },
    GITHUB_CLIENT_SECRET: {
      type: 'string',
    },
  },
}

export async function buildServer() {
  const server = Fastify()

  server.register(fastifyEnv, {
    schema,
    dotenv: {
      path: path.join(__dirname, '../../playground/.env'),
    },
  })
  await server.after()
  server.register(NextAuthPlugin, {
    secret: process.env.NEXTAUTH_SECRET,
    providers: [
      GithubProvider({
        clientId: process.env.GITHUB_CLIENT_ID,
        clientSecret: process.env.GITHUB_CLIENT_SECRET,
      }),
    ],
  } as NextAuthOptions)

  return server
}
