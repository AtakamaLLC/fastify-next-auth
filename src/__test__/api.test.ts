import { afterEach, beforeEach, describe, expect, test } from 'vitest'
import { buildServer } from './buildServer'

let fastify: Awaited<ReturnType<typeof buildServer>>

beforeEach(async () => {
  fastify = await buildServer()
})

afterEach(() => {
  fastify.close()
})

describe('REST API', () => {
  test('GET /api/auth/signin', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/api/auth/signin',
    })

    expect(response.statusCode).toBe(200)
    expect(response.body).toContain('<button type="submit" class="button">Sign in with GitHub</button>')
  })

  test('GET /api/auth/signout', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/api/auth/signout',
    })

    expect(response.statusCode).toBe(200)
    expect(response.body).toContain('Are you sure you want to sign out?')
    expect(response.body).toContain('<button type="submit">Sign out</button>')
  })

  test('GET /api/auth/providers', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/api/auth/providers',
    })

    expect(response.statusCode).toBe(200)
    expect(response.json()).toMatchSnapshot()
  })

  test('GET /api/auth/session', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/api/auth/session',
    })

    expect(response.statusCode).toBe(200)
    expect(response.json()).toMatchSnapshot()
  })

  test('GET /api/auth/csrf', async () => {
    const response = await fastify.inject({
      method: 'GET',
      url: '/api/auth/csrf',
    })

    expect(response.statusCode).toBe(200)
    expect(response.json()).toHaveProperty('csrfToken')
  })
})
