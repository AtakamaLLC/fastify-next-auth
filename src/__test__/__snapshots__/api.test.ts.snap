// Vitest Snapshot v1

exports[`REST API > GET /api/auth/providers 1`] = `
{
  "github": {
    "callbackUrl": "http://localhost:3000/api/auth/callback/github",
    "id": "github",
    "name": "GitHub",
    "signinUrl": "http://localhost:3000/api/auth/signin/github",
    "type": "oauth",
  },
}
`;

exports[`REST API > GET /api/auth/session 1`] = `{}`;
